= AStorage-Java

image:https://img.shields.io/badge/vert.x-4.4.6-white.svg[link="https://vertx.io"]

.Already supported formats (See usage examples below):
* Fasta
* dbNSFP v4.3a
* gnomAD v4
* SpliceAI v1.3
* PharmGKB
* ClinVar
* GTEx v8
* GTF
* GERP
* dbSNP

.Formats mapped in the universal variant query:
* dbNSFP v4.3a
* gnomAD v4
* SpliceAI v1.3
* ClinVar
* GERP
* dbSNP

== Building

To package the application as JAR:
[source]
----
./mvnw clean package
----

== Usage

The service is only running on *localhost:8080* for now. It creates DB directories in the user's home directory if not specified otherwise.

DB directory can be adjusted using a *config.json* file:
[source]
----
{
	"dataDirectoryPath": "/home/user/ExampleStorage"
}
----
To start the application:
[source]
----
java -jar astorage-java-1.0.0.jar [config_json_path]
----
To see the API specification in the Swagger UI (Not complete):
[source]
----
http://localhost:8080/swagger
----

== Notes

* Batch-query parameters match single-query parameters for every format.
* To use the normalization service appropriate genome reference builds(e.g. *GRCh38* and *GRCh37*) should be ingested into *Fasta* first.
* To batch-normalize the data same approach is used as in the batch-query.
* For now the AStorage logs are being written in <dataDirectoryPath>/output.log file.
